engines/nancy/action/sliderpuzzle.cpp

```diff
@@ -0,0 +1,336 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "engines/nancy/action/sliderpuzzle.h"
+
+#include "engines/nancy/util.h"
+#include "engines/nancy/nancy.h"
+#include "engines/nancy/graphics.h"
+#include "engines/nancy/resource.h"
+#include "engines/nancy/sound.h"
+#include "engines/nancy/state/scene.h"
+
+#include "common/serializer.h"
+
+namespace Nancy {
+namespace Action {
+
+Common::Array<Common::Array<int16>> SliderPuzzle::playerTileOrder = Common::Array<Common::Array<int16>>();
```

Comment:
Keep in mind that statics like these must get reset in the ctor of the class - otherwise they might have the wrong state if you restart the engine from the launcher.

Reply:
This particular puzzle keeps its state throughout the entire game, so if the player leaves and comes back the pieces should stay in the same place; resetting them in the constructor would break that. The coming back from the launcher issue gets fixed in Scene::init() where the static _playerHasTriedPuzzle gets reset back to false.

Reply:
Then you have 2 options, either keep its state in the engine class, or create a Singleton. But keeping it like this is not portable.

Reply:
Ah, I did not think of that. Issue should be fixed now, I've removed the static vars from SliderPuzzle and moved them to their own (non-static) struct inside State.